primitives/src/migrate.rs

```diff
@@ -160,14 +181,46 @@ where
         .collect::<Vec<(Vec<u8>, _)>>();
 
     for (key, value) in old_map.into_iter().filter_map(|(raw_key, value)| {
-        let (k1, k2) = decode_double_key::<H, _, _>(&raw_key)?;
+        let (k1, k2) = decode_double_key::<H1, _, H2, _>(&raw_key)?;
         let (kn1, kn2, value) = map(k1, k2, value)?;
-        Some((encode_double_key::<H, _, _>(kn1, kn2), value))
+        Some((encode_double_key::<H1, _, H2, _>(kn1, kn2), value))
     }) {
         put_storage_value(module, item, &key, value);
     }
 }
 
+/// Migrate the values of a double map indexed by keys `K1` and `K2`.
+/// The `map` function transform previous value `V1` into a `V2`.
+/// The double map is located in `module::item`.
+/// `H1` and `H2` are the hashers used with `K1` and `K2` respectively.
+///
+/// It is an optimization to avoid `collect` and it also allows the caller to manage errors during
+/// the migration.
+pub fn migrate_double_map_only_values<'a, V1, V2, H1, K1, H2, K2, F, E>(
+    module: &'a [u8],
+    item: &'a [u8],
+    f: F,
+) -> impl 'a + Iterator<Item = Result<(), MigrationError<E>>>
+where
+    F: 'a + Fn(K1, K2, V1) -> Result<V2, E>,
+    K1: Decode,
+    K2: Decode,
+    H1: ReversibleStorageHasher,
+    H2: ReversibleStorageHasher,
+    V1: 'a + Decode,
+    V2: Encode,
+    E: Clone + Eq + Encode + Decode,
```

Comment:
`Clone + Eq` is never used in the function itself.

Reply:
Required by `MigrationError` (exposed as `Event`).

Reply:
Fixed